File: 42. Trapping Rain Water/trapConstantMemoryUsage.py
from typing import List

class Solution:
  def trap(self, height: List[int]) -> int:
    if not height:
      return 0

    left, right = 0, len(height) - 1
    leftMax, rightMax = height[left], height[right]
    waterTrapped = 0

    while left < right:
      if leftMax < rightMax:
        left += 1
        leftMax = max(leftMax, height[left])
        waterTrapped += leftMax - height[left]
      else:
        right -= 1
        rightMax = max(rightMax, height[right])
        waterTrapped += rightMax - height[right]

    return waterTrapped

solution = Solution()
height = [0,1,0,2,1,0,1,3,2,1,2,1]
print('trap: ', solution.trap(height))

File: 42. Trapping Rain Water/trapConstantMemoryUsage.js
/**
 * @param {number[]} height
 * @return {number}
 */
const trap = (height) => {
  if (!height || !height.length) {
    return 0;
  }

  let left = 0;
  let right = height.length - 1;
  let maxLeft = height[left];
  let maxRight = height[right];
  let waterTrapped = 0;

  while (left < right) {
    if (maxLeft < maxRight) {
      left += 1;
      maxLeft = Math.max(maxLeft, height[left]);
      waterTrapped += maxLeft - height[left];
    } else {
      right -= 1;
      maxRight = Math.max(maxRight, height[right]);
      waterTrapped += maxRight - height[right];
    }
  }

  return waterTrapped;
};

height = [0,1,0,2,1,0,1,3,2,1,2,1]
console.log('trap: ', trap(height))
